File: Python/binaryInsertionSort.py
def binary_search(arr, val, start, end):

    while start <= end:
        mid = (start + end) // 2  # Calculate the mid index

        # If the value is found at mid, return mid
        if arr[mid] == val:
            return mid
        
        # If the value is less than the middle element, search the left half
        elif arr[mid] > val:
            end = mid - 1
        
        # If the value is greater than the middle element, search the right half
        else:
            start = mid + 1

    return start  # The position where val should be inserted

def binary_insertion_sort(arr):
    
    # Traverse from the second element to the end of the array
    for i in range(1, len(arr)):
        key = arr[i]  # Current value to be inserted
        # Find the position where key should be inserted
        j = binary_search(arr, key, 0, i - 1)

        # Shift elements to the right to make space for the key
        arr = arr[:j] + [key] + arr[j:i] + arr[i + 1:]

    return arr  # Return the sorted array

# Example usage
if __name__ == "__main__":
    # Sample array to sort
    sample_array = [5, 2, 9, 1, 5, 6]
    print("Original array:", sample_array)
    
    # Sorting the array
    sorted_array = binary_insertion_sort(sample_array)
    print("Sorted array:", sorted_array)

File: C++/binaryInsertionSort.cpp
#include <iostream>
#include <vector>

// Function for binary search to find the correct position to insert an element
// We search in the range from 'low' to 'high'
int binarySearch(const std::vector<int>& arr, int item, int low, int high) {
    // Loop to repeatedly narrow the search range
    while (low <= high) {
        // Calculate the middle index of the current search range
        int mid = low + (high - low) / 2;

        // If the item is equal to the middle element, we return the next position
        if (item == arr[mid]) {
            return mid + 1; // Place it right after the same element
        }

        // If the item is greater than the middle element, it belongs to the right half
        if (item > arr[mid]) {
            low = mid + 1; // Adjust the lower bound
        } else {
            // Otherwise, it belongs to the left half
            high = mid - 1; // Adjust the upper bound
        }
    }
    // If not found, return the position where it should be inserted (in sorted order)
    return low;
}

// Function that implements the binary insertion sort algorithm
void binaryInsertionSort(std::vector<int>& arr) {
    int n = arr.size(); // Get the size of the array

    // Start from the second element (index 1) and move through the array
    for (int i = 1; i < n; ++i) {
        int current = arr[i]; // Store the current element to be inserted
        int j = i - 1; // j tracks the sorted portion of the array (on the left)

        // Use binary search to find the position where 'current' should be inserted
        int pos = binarySearch(arr, current, 0, j);

        // Shift all elements that are greater than 'current' one position to the right
        while (j >= pos) {
            arr[j + 1] = arr[j]; // Shift element to the right
            j--; // Move left to check the next element
        }

        // Insert 'current' into its correct position
        arr[pos] = current;
    }
}

// Helper function to print the elements of the array
void printArray(const std::vector<int>& arr) {
    for (int i = 0; i < arr.size(); ++i) {
        std::cout << arr[i] << " "; // Print each element followed by a space
    }
    std::cout << std::endl; // Print a newline at the end
}

// Main function
int main() {
    // Example array to sort
    std::vector<int> arr = {37, 23, 17, 5, 88, 61, 9};

    // Output the original array
    std::cout << "Original array: ";
    printArray(arr);

    // Call the binary insertion sort function
    binaryInsertionSort(arr);

    // Output the sorted array
    std::cout << "Sorted array: ";
    printArray(arr);

    return 0;
}
